### ISSUE TITLE
Support the official MySQL Connector/Python

### ISSUE THREAD

**Issue by jamadden** [zodb/relstorage#155]
As distributed by Oracle: http://dev.mysql.com/doc/connector-python/en/

It has an optional C extension.

It's unfortunately hard-ish to install since it doesn't officially release to PyPI.

It has the advantage of [automatically supporting a single prepared statement per cursor](http://dev.mysql.com/doc/connector-python/en/connector-python-api-mysqlconnection-cursor.html) *with* parameters using the binary protocol, which means there's not a server round-trip to set user variables for each parameter. (Unfortunately calling stored procs still uses user variables.) This may or may not be useful to us. (Update: It probable would be, if we can make it work. It improved performance by 40% on postgres.)

Other advantages not in any other driver include automatic connection pooling (not especially useful since the higher level caches will not be present), automatic failover, and support for [MySQL Fabric](https://dev.mysql.com/downloads/fabric/)

**Comment by jamadden** [comment]
Initial impressions:

- It's working on Python 3, but [has issues on Python 2](https://travis-ci.org/zodb/relstorage/builds/195001868). At least one issue is that it returns data as `bytearray` instead of byte. I don't know if that can be configured.
- It has a C extension, but even when it's installed it's not used by default.
- It's slow, at least without the C extension. Compare 4:25 with PyMySQL vs 7:15 on Python 3.6.
- Trying to use the native prepared statement support is going to be tricky. I've only done some early experiments, but it looks like you have to completely give up on the C extension, there are lots of restrictions (it won't gracefully fall back to plain statements for things that can't be prepared), and the multiple Cursor subclasses needed to get buffering and prepared statements may not play well together.

I'm not ready to give up, but it's not looking great.

**Comment by jamadden** [comment]
Its looking substantially better today. Travis is green for Python 2/3/pypy with C and python implementations. Commit c5200e8138e740b1cee79f1059bf761b0868f651 brings performance up to par with the other drivers (at least on my machine). 

This is a stricter driver and has revealed some broken assumptions in the relstorage code, so that's good.

**Comment by jamadden** [comment]
Work is in https://github.com/zodb/relstorage/compare/mysql-connector

**Comment by jamadden** [comment]
We're finally at the point we can test the native binary prepared statement support. I did just a little bit of work to be able to use it just for `load_current`, the single most executed query taking a parameter. 

The results are disappointing.

Over a Unix socket, using `load_current` as a prepared statement, the zodbshootout for warm and cold are just about the same (7723/7619) as for *not* using the prepared statement (7866/7729). In fact they're a bit lower, but within the margins. Both numbers are lower than what PyMySQL produces (8504/8215) (remember that prepared statements only work with the python implementation).

Over TCP (to localhost), things are worse. We expect things to be slower due to communication overhead, and they are. PyMySQL delivers 7316/7181. But the prepared statement version gives 5994/6046 compared to 6679/6587 without. 

Digging into this, I find that there is *still* an extra packet sent: before the prepared statement is executed, the driver sends [COM_STMT_RESET](https://dev.mysql.com/doc/internals/en/com-stmt-reset.html). This is necessary if you stream long binary data to the prepared statement, or have it allocate a cursor. 

We do neither, so I commented that out. The numbers went up a bit (6061/6016), but still not enough to match or better the immediate version. 

Profiling revealed no obvious hot spots for optimization.

My conclusions are:

1. For a SELECT statement this simple, the overhead of parsing and planning for MySQL is negligible. So prepared statements gain us little or nothing.
2. The management overhead of the prepared statement support in MySQL Connector/Python actually hurts us, small though it is. If the statement were more complex, we might have more benefit.

(All tests were done with history free databases.)
